src/coreclr/jit/importercalls.cpp

```diff
@@ -6730,6 +6730,21 @@ void Compiler::impMarkInlineCandidate(GenTree*               callNode,
     // candidate, we're done.
     if (call->IsInlineCandidate() || !call->IsGuardedDevirtualizationCandidate())
     {
+        // Runtime lookup is expected to be spilled into a temp for inline candidates.
+        CallArg* instParam = call->gtArgs.FindWellKnownArg(WellKnownArg::InstParam);
+        if ((instParam != nullptr) && instParam->GetNode()->OperIs(GT_RUNTIMELOOKUP))
```

Comment:
The `OperIs(GT_RUNTIMELOOKUP)` check is odd to me. Can the check be `!OperIs(GT_LCL_VAR)` instead, to make the invariants more clear?

Reply:
Why would we want to spill GT_CNS_INT context? or any other non-RUNTIMELOOKUP one

Reply:
Ok, I think I misunderstood. So we will only call `getRuntimeContextTree` in the inlinee if we in the caller needed a runtime lookup for the instantiation arg?

It seems it would be overall better to make the handling uniform with how other inline arguments are handled to avoid the complicated contract here. `impInlineFetchArg` has a lot of smart handling in it, so maybe that would help with CQ as well (e.g. avoid the spilling in some common cases, like if the generic context ends up unused in the inlinee). The spilling would be handled only when necessary by `fgInlinePrependStatements`.

Reply:
Basically the idea would be:
1. Create a fully fledged `InlArgInfo` for the context argument. Store it separately from `inlArgInfo` in `InlineInfo`.
2. Refactor `impInlineFetchArg` to avoid needing `lclNum` (just requires passing the `InlArgInfo` and `InlLclVarInfo` entries directly)
3. Switch `getRuntimeContextTree` to use `impInlineFetchArg` on the `InlArgInfo` created for the context argument
4. Add handling for the new `InlArgInfo` to `fgInlinePrependStatements`, to make sure it's spilled if used or completely removed if unused.

Could be done in a separate PR...

Reply:
@jakobbotsch good idea! Addressed